File: 02-dive-2.hs
module Dive where

import System.IO (IOMode(ReadMode), openFile, hGetContents)

type Forward = Int
type Depth = Int
type Aim = Int
type Movement = (Forward, Depth, Aim)
type Command = (String, Int)
type Line = String

move ::  Movement -> Command -> Movement
move (f,d,a) ("down", x)    = (f, d, a + x)
move (f,d,a) ("up", x)      = (f, d, a - x)
move (f,d,a) ("forward", x) = (f + x, d + a * x, a)

process :: Movement -> Line -> Movement
process m = move m . asCommand . words

asCommand :: [String] -> Command
asCommand (s:i:_) = (s, read i::Int)

answer :: Movement -> Int
answer (f, d, _) = f * d

main = do  
  let list = []
  handle <- openFile "02-input.txt" ReadMode
  contents <- hGetContents handle

  print $ answer $ foldl process (0,0,0) $ lines contents

File: 02-dive-2.rs
use std::fs::File;
use std::io::{BufRead, BufReader};

fn main() {
    let mut forward:i32=0;
    let mut depth:i32=0;
    let mut aim:i32=0;

    let file = File::open("./02-input.txt").unwrap();
    let reader = BufReader::new(file);
    for (_index, line) in reader.lines().enumerate() {
        let line = line.unwrap();
        let parts = line.split_once(" ").map(|(s,i)| (s, i.parse::<i32>()));

        if let Some((command, Ok(x))) = parts {
            match command {
                "down" =>
                    aim = aim + x,
                "up" => 
                    aim = aim - x,
                "forward"=> {
                    forward = forward + x;
                    depth = depth + (aim * x);
                },
                _ => (),
            }
        }
    }

    println!("{}", forward * depth)
}

